Compare the text and questions input by LINE webhook.

LittleYenMin opened the pull request:
#11 

The data format of the result of text similarity.
```json
{
   "query": "Book me a flight to Cairo",
   "topScoringIntent": {
       "intent": "BookFlight",
       "score": 0.9887482
   },
   "intents": [
       {
           "intent": "BookFlight",
           "score": 0.9887482
       },
       {
           "intent": "None",
           "score": 0.04272597
       },
       {
           "intent": "LocationFinder",
           "score": 0.0125702191
       },
       {
           "intent": "Reminder",
           "score": 0.00375502417
       },
       {
           "intent": "FoodOrder",
           "score": 3.765154E-07
       },
   ]
}
```

LittleYenMin committed:
add: method for compare input with question

LittleYenMin committed:
add: Intent DTO.

LittleYenMin committed:
add: method to convert the type of  dict of intents to object.

LittleYenMin committed:
update: to more cleaner doc-test.

LittleYenMin committed:
add: repr to intent object (return string of value of object).

LittleYenMin committed:
add: doc-test for questions method, to test convert the type of dict of intent  to array of intent object.

LittleYenMin committed:
update: better test data for doc-test of data/compare.

LittleYenMin committed:
add: QueryResult object of data compare with nlu.similarity

LittleYenMin committed:
update: assign value to top-scoring-intent when it is None.

LittleYenMin committed:
add: __repr__ of QueryResult

LittleYenMin committed:
update: return QueryResult to instead array of intents.

LittleYenMin committed:
update: doc-test for return value changed.

LittleYenMin committed:
format: autopep8.

LittleYenMin committed:
update: compare user input similarity with data.csv by LINE SDK.

LittleYenMin commented on app.py:
```diff
@@ -37,10 +38,11 @@ def callback():
 def handle_message(event):
-    text = nlu.response_message(event.message.text)
+    result = data.compare.questions(word=event.message.text, questions=data.source.from_csv_file('./data.csv'))
+    json_str = json.dumps(result, default=lambda o: o.__dict__, ensure_ascii=False)
```
[python json.dumps 中的ensure_ascii 参数引起的中文编码问题](https://www.cnblogs.com/biangbiang/archive/2013/02/19/2916780.html)

LittleYenMin replied:
Example:

set ensure_ascii with default True
```json
{"query": "\u6210\u7e3e", "intents": [{"intent": "\u660c\u660c;\u66e0\u8ab2", "score": 0.0}, {"intent": "\u660c\u660c;\u8ab2\u8868", "score": 0.0}, {"intent": "\u660c\u660c;\u6210\u7e3e", "score": 0.0}], "topScoringIntent": {"intent": "\u660c\u660c;\u66e0\u8ab2", "score": 0.0}}
```

set ensure_ascii to False
```json
{"query": "成績", "intents": [{"intent": "昌昌;曠課", "score": 0.0}, {"intent": "昌昌;課表", "score": 0.0}, {"intent": "昌昌;成績", "score": 0.0}], "topScoringIntent": {"intent": "昌昌;曠課", "score": 0.0}}
```

LittleYenMin replied:
@dannypsnl 
Which one is better ? Extract the JSON dumps to QuestionResult or keep intact?

extract to QuestionResult
```python
import json


class QueryResult(object):
    ...

    def json(self):
        return json.dumps(self, default=lambda o: o.__dict__, ensure_ascii=False)
```

dannypsnl replied:
Extract method at most time is a good idea, but `to_json(self)` would be better.

LittleYenMin committed:
rename: questions to examples.

LittleYenMin committed:
rename: examples to samples.

LittleYenMin committed:
format: comment.

dannypsnl commented on data/compare.py:
```diff
@@ -0,0 +1,58 @@
+    >>> questions('question-1-3', [ExampleQuestion('question-1', 'answer-A'), ExampleQuestion('question-2', 'answer-B')])
+    <QueryResult query question-1-3 topScoringIntent <Intent answer-A 0.8728715609439696> intents [<Intent answer-A 0.8728715609439696>, <Intent answer-B 0.6546536707079772>]>
+    """
+    intents = _questions(word, samples)
```
I don't understand the dispatching here, `_questions` give you nothing except confusing.

dannypsnl commented on data/compare.py:
```diff
@@ -0,0 +1,58 @@
+    <QueryResult query question-1-3 topScoringIntent <Intent answer-A 0.8728715609439696> intents [<Intent answer-A 0.8728715609439696>, <Intent answer-B 0.6546536707079772>]>
+    """
+    intents = _questions(word, samples)
+    result = [Intent(intent=intent, score=score)
```
And the conversion seems could be done by setup place, why need to convert **map** to **list**?

dannypsnl commented on data/compare.py:
```diff
@@ -0,0 +1,58 @@
+    for sample in samples:
+        command = sample.command_type
+        similarity = nlu.similarity(sample.question, word)
+        intents[command] = max(similarity, intents.get(command, 0))
```
If I understand the code correctly, you can have one function by following code:
```py
intents = []
for sample in samples:
    command = sample.command_type
    similarity = nlu.similarity(sample.question, word)
    intents.append(Intent(command, max(similarity, intents.get(command, 0))))
```

dannypsnl commented on data/compare.py:
```diff
@@ -0,0 +1,58 @@
+
+def questions(word: str, samples: [ExampleQuestion]) -> QueryResult:
+    """
+    >>> questions('question-1-3', [ExampleQuestion('question-1', 'answer-A'), ExampleQuestion('question-2', 'answer-B')])
```
As I said, make your test explain the functionality of the function. It must have a meaningful example to do that, else remove the missing test here.

dannypsnl commented on data/compare.py:
```diff
@@ -0,0 +1,58 @@
+
+def _questions(word: str, samples: [ExampleQuestion]) -> dict:
+    """
+    >>> _questions('question-1-3', [ExampleQuestion('question-1', 'answer-A'), ExampleQuestion('question-2', 'answer-B')])
```
And here, we don't need to test underlying code if you already test higher level part, unless the situation is the algorithm is too complicated, but not here, at this place your higher level function just convert the **map** to **list**, seems it won't make test more complicated, why we have the same test here?

dannypsnl commented on data/compare.py:
```diff
@@ -0,0 +1,58 @@
+    return QueryResult(query_text=word, intents=result)
+
+
+def _questions(word: str, samples: [ExampleQuestion]) -> dict:
```
`List[T]`
`-> Dict[keyT, valueT]`

dannypsnl commented on data/compare.py:
```diff
@@ -0,0 +1,58 @@
+            intent=self.intent, score=self.score)
+
+
+class QueryResult(object):
```
What kind of query? What is the result?

LittleYenMin replied:
It's a JSON object of the result more like the DTO.

```json
{
   "query": "Book me a flight to Cairo",
   "topScoringIntent": {
       "intent": "BookFlight",
       "score": 0.9887482
   },
   "intents": [
       {
           "intent": "BookFlight",
           "score": 0.9887482
       },
       {
           "intent": "None",
           "score": 0.04272597
       },
       {
           "intent": "LocationFinder",
           "score": 0.0125702191
       },
       {
           "intent": "Reminder",
           "score": 0.00375502417
       },
       {
           "intent": "FoodOrder",
           "score": 3.765154E-07
       },
   ]
}
```

LittleYenMin replied:
might be renamed to `SimilarityResult` would be better.

dannypsnl commented on data/compare.py:
```diff
@@ -0,0 +1,58 @@
+from questions import ExampleQuestion
+
+
+class Intent(object):
```
Intent what?

LittleYenMin committed:
update: return type of questions, return array of intent to instead the type of dict.

LittleYenMin committed:
update: doc-test for _questions.

LittleYenMin committed:
remove: useless doc-test for list to DTO.

LittleYenMin committed:
update: fancy way to set topScoringIntent by getter and setter.

LittleYenMin commented on data/compare.py:
```diff
@@ -0,0 +1,58 @@
+            if top_scoring_intent is not None:
+                top_scoring_intent = intent if intent.score > top_scoring_intent.score else top_scoring_intent
+            else:
+                top_scoring_intent = intent
```
yep, you're right.

ref: [Get the object with the max attribute value in a list of objects](https://stackoverflow.com/questions/24209255/get-the-object-with-the-max-attribute-value-in-a-list-of-objects)

LittleYenMin committed:
Rename: QueryResult to SimilarityResult.

LittleYenMin committed:
update: extract json dumps to method of SimilarityResult named to_json

LittleYenMin committed:
Rename: Intent to Similarity.

LittleYenMin committed:
merge: data.compare and data.source to data.py

LittleYenMin committed:
remove: useless import.

LittleYenMin committed:
format: autopep8

LittleYenMin committed:
update: Readme for real-usage image.

LittleYenMin committed:
update: misspell.

LittleYenMin committed:
update: __repr__ cause by class renamed but forgot the method.

dannypsnl commented on data.py:
```diff
@@ -0,0 +1,83 @@
+    def intents(self, intents: [Similarity]):
+        self._intents = intents
+        self.topScoringIntent = max(
+            self.intents, key=operator.attrgetter('score'))
```
I don't like this approach; remember, `'score'` is not `.score`, the difference is you can't get any hint from a string, but you could have one originally, at here we have a common way is:
```py
max(list, key=lambda item: item.score)
```

dannypsnl commented on data.py:
```diff
@@ -0,0 +1,83 @@
+    def to_json(self):
+        return json.dumps(
+            self,
+            default=lambda o: o.__dict__,
```
I would say remove the setter magic, what's the problem of `set_intent`?

Don't you think directly accessing something is start with `__` must be something wrong?

dannypsnl replied:
I guess I know why you use the getter/setter, but I said, the cache won't perform as good as you thought! At the most time, we use it once, if you never use it, then caching can be even worse.

dannypsnl commented on data.py:
```diff
@@ -0,0 +1,83 @@
+
+
+def from_csv_file(path: str) -> [questions.ExampleQuestion]:
+    with open(path, 'r', encoding='utf-8') as f:
```
I'm thinking why we do not return `open(path, 'r', encoding='utf-8')`? Any ideas? Seems `with` is no need.

LittleYenMin replied:
without `with` means we need to close the file manually at the end of function.

dannypsnl replied:
https://stackoverflow.com/questions/1832528/is-close-necessary-when-using-iterator-on-a-python-file-object

According to the answer, GC would collect it, but `with` is a better practice. So we keep it.

dannypsnl commented on data.py:
```diff
@@ -0,0 +1,83 @@
+
+def compare_questions(
+    word: str, samples: [
+        questions.ExampleQuestion]) -> SimilarityResult:
```
The indent is weird, why there's a newline?

LittleYenMin replied:
I don't know, it's created by autopep8.
How about I put it back to the same line.

```python
def compare_questions(word: str, samples: [questions.ExampleQuestion]) -> SimilarityResult:
```

dannypsnl replied:
`List[questions.ExampleQuestion]`

dannypsnl commented on data.py:
```diff
@@ -0,0 +1,83 @@
+    return SimilarityResult(query_text=word, intents=intents)
+
+
+def _questions(
```
The function is useless; why not move into `compare_questions`? If you prefer to keep it, then the naming is missing, it called `_questions` but returns `intents`.

LittleYenMin replied:
ummmmmm, you're right.

dannypsnl commented on data.py:
```diff
@@ -0,0 +1,83 @@
+def compare_questions(
+    word: str, samples: [
+        questions.ExampleQuestion]) -> SimilarityResult:
+    intents = _questions(word, samples)
```
Remove the temporary variable is a good start.

dannypsnl commented on data.py:
```diff
@@ -0,0 +1,83 @@
+            intent=self.intent, score=self.score)
+
+
+class SimilarityResult(object):
```
It's more like a report; later we base on this report to pick a command, what do you think?

dannypsnl commented on app.py:
```diff
@@ -36,12 +36,12 @@ def callback():
 
 @handler.add(MessageEvent, message=TextMessage)
 def handle_message(event):
-    similarty_result = data.compare_questions(
```
You should have a spell checking plugin in your editor XD.

dannypsnl commented:
@LittleYenMin I don't like the branch commit message with type now, that should use only for PR commit as we do final, to explain target of the whole this stuff. 

If you add a type for every commit, you will find that's hard to pick a correct type for it because of the changed is too small for the most time. And you would lose some space to explain the commit better, so why not describe the commit better, remove the type of it?

LittleYenMin committed:
custom setter to instead to original magic getter/setter.
- update: remove original magic setter
- update: access score by attribute not string.
